csrc/scheduler/normalization_utils.cpp

```diff
@@ -1235,14 +1236,32 @@ void movePersistentBufferToSmem(
       // path of gmem -> smem to reduce temporary register usage. Otherwise, the
       // data path from gmem to shared memory (smem) follows this sequence: gmem
       // -> L1 cache -> register -> smem.
+      // Only supported after device 8.0 and requires vectorized load.
       int hw_major = at::cuda::getCurrentDeviceProperties()->major;
-      if (is_cached_input && hw_major >= 8) {
+      if (rparams->vectorize_inner_reduction && is_cached_input &&
+          hw_major >= 8) {
         tv->definition()->as<LoadStoreOp>()->setOpType(
             LoadStoreOpType::CpAsync);
         tv->definition()->as<LoadStoreOp>()->setCacheOp(CacheOp::Unspecified);
       }
+      // T4_s_float = CpAsync(T0_g_float)
+      // do a register cache for all the uses of this smem tv
+      auto cached_tv = tv->cacheAfter();
+      const auto& consumers = ir_utils::consumerTvsOf(cached_tv);
+      smem_consumers.push_back(cached_tv);
+      for (auto i = 1; i < (int)consumers.size(); i++) {
+        auto consumer = consumers.at(i);
+        // recompute cached_tv for each consumer, so it is no longer persistent
+        // similar to project to inputs, here we are projecting to the shared
+        // memory buffer.
+        auto cached_tv_replicate = RecomputeTv::recompute(cached_tv, {tv});
+        ir_utils::replaceValInExprInputs(
+            consumer->definition(), cached_tv, cached_tv_replicate);
+        smem_consumers.push_back(cached_tv_replicate);
+      }
```

Comment:
Why does this avoid bank conflicts?

Reply:
I'm sorry my question wasn't clear. Why does the use of the shmem buffer need to be replicated?

Reply:
The `smem_tv` is a cached input and is a persistent buffer, after doing `cached_tv = cacheAfter(smem_tv)`, the `cached_tv` becomes the new persistent buffer and it is a register tv, however, we want to avoiding using registers, so we need to recompute all the uses of `cached_tv` except the first uses. Similar to project to inputs, here we are projecting to the `smem_tv`. 
Without this `recompute`, `T6_l_float` in the fusion_ir is also a persistent buffer.
```
Inputs:
  T0_g_float[ iS0{i0}, iS73{( ceilDiv(( ceilDiv(( ceilDiv(( i2 * ( i3 * i4 ) ), 4) ), 9) ), 1) )}, iS70{9}, iS72{1}, iS69{4} ]
Outputs:
  T3_g_float[ iblockIdx.x20{i0}, ithreadIdx.x97{( ceilDiv(( ceilDiv(( ceilDiv(( i2 * ( i3 * i4 ) ), 4) ), 9) ), 1) )}_p, iS94{9}, iUS96{1}, iV93{4} ] ca_pos( 4 ) produce_pos( 4 )

%kernel_math {
T4_s_float[ iblockIdx.x16{i0}, ithreadIdx.x65{( ceilDiv(( ceilDiv(( ceilDiv(( i2 * ( i3 * i4 ) ), 4) ), 9) ), 1) )}_p, iS62{9}, iUS64{1}, iV61{4} ] ca_pos( 4 )
   = CpAsync( T0_g_float[ iS0{i0}, iS73{( ceilDiv(( ceilDiv(( ceilDiv(( i2 * ( i3 * i4 ) ), 4) ), 9) ), 1) )}, iS70{9}, iS72{1}, iS69{4} ] )
T6_l_float[ iblockIdx.x24{i0}, ithreadIdx.x57{( ceilDiv(( ceilDiv(( ceilDiv(( i2 * ( i3 * i4 ) ), 4) ), 9) ), 1) )}_p, iS54{9}, iUS56{1}, iV53{4} ] ca_pos( 1 ) produce_pos( 4 )
   = Set( T4_s_float[ iblockIdx.x16{i0}, ithreadIdx.x65{( ceilDiv(( ceilDiv(( ceilDiv(( i2 * ( i3 * i4 ) ), 4) ), 9) ), 1) )}_p, iS62{9}, iUS64{1}, iV61{4} ] ca_pos( 4 ), cache_op=Streaming )
T7_l_float[ iblockIdx.x36{i0}, ithreadIdx.x47{( ceilDiv(( ceilDiv(( ceilDiv(( i2 * ( i3 * i4 ) ), 4) ), 9) ), 1) )}rf_p, rS44{9}rf, rUS46{1}rf, rS43{4}rf ] ca_pos( 2 ) produce_pos( 1 )
   = reduction( T6_l_float[ iblockIdx.x24{i0}, ithreadIdx.x57{( ceilDiv(( ceilDiv(( ceilDiv(( i2 * ( i3 * i4 ) ), 4) ), 9) ), 1) )}_p, iS54{9}, iUS56{1}, iV53{4} ] ca_pos( 1 ) produce_pos( 4 ), op = add, initial value = float(0), allreduce = false )
T1_l_float[ iblockIdx.x48{i0}, rthreadIdx.x49{( ceilDiv(( ceilDiv(( ceilDiv(( i2 * ( i3 * i4 ) ), 4) ), 9) ), 1) )}_p ] ca_pos( 1 ) produce_pos( 2 )
   = reduction( T7_l_float[ iblockIdx.x36{i0}, ithreadIdx.x47{( ceilDiv(( ceilDiv(( ceilDiv(( i2 * ( i3 * i4 ) ), 4) ), 9) ), 1) )}rf_p, rS44{9}rf, rUS46{1}rf, rS43{4}rf ] ca_pos( 2 ) produce_pos( 1 ), op = add, initial value = float(0), allreduce = false )
T2_l_float[ iblockIdx.x8{i0}, bthreadIdx.x89{( ceilDiv(( ceilDiv(( ceilDiv(( 1 * ( 1 * 1 ) ), 4) ), 9) ), 1) )}_p, bS86{9}, bUS88{1}, bS85{4} ] ca_pos( 1 ) produce_pos( 1 )
   = broadcast( T1_l_float[ iblockIdx.x48{i0}, rthreadIdx.x49{( ceilDiv(( ceilDiv(( ceilDiv(( i2 * ( i3 * i4 ) ), 4) ), 9) ), 1) )}_p ] ca_pos( 1 ) produce_pos( 2 ) )
T5_l_float[ iblockIdx.x12{i0}, ithreadIdx.x81{( ceilDiv(( ceilDiv(( ceilDiv(( i2 * ( i3 * i4 ) ), 4) ), 9) ), 1) )}_p, iS78{9}, iUS80{1}, iS77{4} ] ca_pos( 4 ) produce_pos( 1 )
   = T6_l_float[ iblockIdx.x24{i0}, ithreadIdx.x57{( ceilDiv(( ceilDiv(( ceilDiv(( i2 * ( i3 * i4 ) ), 4) ), 9) ), 1) )}_p, iS54{9}, iUS56{1}, iV53{4} ] ca_pos( 1 ) produce_pos( 4 )
   / T2_l_float[ iblockIdx.x8{i0}, bthreadIdx.x89{( ceilDiv(( ceilDiv(( ceilDiv(( 1 * ( 1 * 1 ) ), 4) ), 9) ), 1) )}_p, bS86{9}, bUS88{1}, bS85{4} ] ca_pos( 1 ) produce_pos( 1 );
T3_g_float[ iblockIdx.x20{i0}, ithreadIdx.x97{( ceilDiv(( ceilDiv(( ceilDiv(( i2 * ( i3 * i4 ) ), 4) ), 9) ), 1) )}_p, iS94{9}, iUS96{1}, iV93{4} ] ca_pos( 4 ) produce_pos( 4 )
   = Set( T5_l_float[ iblockIdx.x12{i0}, ithreadIdx.x81{( ceilDiv(( ceilDiv(( ceilDiv(( i2 * ( i3 * i4 ) ), 4) ), 9) ), 1) )}_p, iS78{9}, iUS80{1}, iS77{4} ] ca_pos( 4 ) produce_pos( 1 ), cache_op=Streaming )
} // %kernel_math 
```
The code is:
```
  Array<float, 36LL, 4> T6;
  #pragma unroll
  for(nvfuser_index_t i10 = 0LL; i10 < 9LL; ++i10) {
    T6.set(float(0.000000000e+00f));
  }
  NVFUSER_UPDATE_MAGIC_ZERO;
  #pragma unroll
  for(nvfuser_index_t i10 = 0LL; i10 < 9LL; ++i10) {
    nvfuser_index_t i11;
    i11 = i5 * (i10 + nvfuser_zero);
    if ((b8 && (i9 < (-i11)))) {
      asm volatile(
        "{\n"
        "  .reg .pred p0; \n"
        "  setp.ne.b32 p0, %3, 0;\n"
        "  cp.async.ca.shared.global [%0], [%1], %2, p0;\n"
        "}\n"
        :
        :"r"((uint32_t)(i6)),
         "l"((ptr2 + i11)),
         "n"(16LL),
         "n"((uint32_t)(false))
      );
      asm volatile("cp.async.wait_all;\n");
      loadGeneric<float, 4>( &T6[(4LL * i10)],  &T4[i0]);
    }
  }
```

Reply:
OK, got it. Let me add some comment suggestions.